In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import country_converter as coco
import warnings
warnings.filterwarnings('ignore')

In [2]:
original_df = pd.read_csv("data/stackoverflow_users_2016-2019-notext.csv")

In [3]:
"""Show the data."""
display(original_df[100:120])

id  age                             location  \
100  6163527  NaN                                  NaN   
101  6169583  NaN            Dallas, TX, United States   
102  6175679  NaN  Perth, Western Australia, Australia   
103  6183039  NaN                              Nigeria   
104  6184739  NaN                         Tokyo, Japan   
105  6190338  NaN                 Bentonville, AR, USA   
106  6195678  NaN             Saint Petersburg, Россия   
107  6205655  NaN                                  NaN   
108  6223963  NaN                            Sri Lanka   
109  6231885  NaN                                  NaN   
110  6237339  NaN                       Düren, Germany   
111  6244279  NaN        Buenos Aires, CABA, Argentina   
112  6248181  NaN          New York, NY, United States   
113  6259321  NaN                 México, CDMX, México   
114  6259476  NaN                                  NaN   
115  6261859  NaN                                  NaN   
116  6262233  NaN                                  NaN   
117  6268538  NaN           Houston, TX, United States   
118  6275487  NaN                          Netherlands   
119  6277964  NaN      Ghaziabad, Uttar Pradesh, India   

                   creation_date  reputation  up_votes  down_votes  
100  2016-04-05 20:54:21.173 UTC         136         4           2  
101  2016-04-07 00:25:48.483 UTC         181         1           0  
102  2016-04-08 06:12:56.763 UTC         717       124           2  
103   2016-04-10 04:17:02.92 UTC         138        25           0  
104  2016-04-10 16:26:57.327 UTC         242       114           9  
105   2016-04-11 21:19:45.63 UTC         448       494           7  
106  2016-04-12 21:34:53.497 UTC         841        60           5  
107  2016-04-14 18:17:02.893 UTC         367         7           1  
108  2016-04-19 09:11:15.923 UTC          82        76           0  
109  2016-04-20 18:21:37.523 UTC         123         6           0  
110   2016-04-21 19:01:00.99 UTC         688        49          16  
111  2016-04-23 11:22:35.523 UTC         114        13           0  
112  2016-04-24 17:18:49.303 UTC          99         7           0  
113  2016-04-27 02:55:27.217 UTC          55        16           0  
114  2016-04-27 04:02:35.813 UTC          30        57           0  
115  2016-04-27 13:55:11.643 UTC          17         0           0  
116   2016-04-27 15:10:35.56 UTC         324         1           0  
117  2016-04-28 19:51:59.827 UTC         103         7           0  
118   2016-04-30 15:48:09.21 UTC         152        78           3  
119   2016-05-01 13:57:49.31 UTC          43         5           0

The location field seems to have the country name as the last part quite consistently, separated 
by a comma and space. Some of the countries are, however, named using some non-English form. 
Let's download a dataset of country names in different languages and use them to match locations to countries.

In [4]:
"""Extract the country field out of the location field"""
original_df["country"] = original_df["location"].apply(lambda x: x.split(", ")[-1].lower() if isinstance(x,str) else np.nan)

In [5]:
original_df.head(5)

id  age             location                creation_date  reputation  \
0  5749314  NaN       Moscow, Russia   2016-01-05 19:30:46.89 UTC         628   
1  5751170  NaN           Bangladesh  2016-01-06 07:51:18.563 UTC         412   
2  5752638  NaN  Berlin, Deutschland   2016-01-06 14:11:34.04 UTC          58   
3  5754033  NaN              Ukraine   2016-01-06 19:20:31.56 UTC          44   
4  5757173  NaN                  NaN  2016-01-07 11:20:00.433 UTC          12   

   up_votes  down_votes      country  
0        34          69       russia  
1        11           2   bangladesh  
2         4           0  deutschland  
3        96           0      ukraine  
4         0           0          NaN

In [6]:
"""Load the dataset of country names in different languages got from ip2location.com."""
country_names = pd.read_csv("data/IP2LOCATION-COUNTRY-MULTILINGUAL/IP2LOCATION-COUNTRY-MULTILINGUAL.CSV")

In [7]:
display(country_names)

LANG            LANG_NAME COUNTRY_ALPHA2_CODE COUNTRY_ALPHA3_CODE  \
0         AF            AFRIKAANS                  AD                 AND   
1         AF            AFRIKAANS                  AE                 ARE   
2         AF            AFRIKAANS                  AF                 AFG   
3         AF            AFRIKAANS                  AG                 ATG   
4         AF            AFRIKAANS                  AI                 AIA   
...      ...                  ...                 ...                 ...   
20164  ZH-TW  CHINESE TRADITIONAL                  YE                 YEM   
20165  ZH-TW  CHINESE TRADITIONAL                  YT                 MYT   
20166  ZH-TW  CHINESE TRADITIONAL                  ZA                 ZAF   
20167  ZH-TW  CHINESE TRADITIONAL                  ZM                 ZMB   
20168  ZH-TW  CHINESE TRADITIONAL                  ZW                 ZWE   

       COUNTRY_NUMERIC_CODE                COUNTRY_NAME  
0                        20                     Andorra  
1                       784  Verenigde Arabiese Emirate  
2                         4                  Afganistan  
3                        28          Antigua en Barbuda  
4                       660                    Anguilla  
...                     ...                         ...  
20164                   887                          也門  
20165                   175                         馬約特  
20166                   710                          南非  
20167                   894                         贊比亞  
20168                   716                        津巴布韋  

[20169 rows x 6 columns]

In [8]:
"""Pick only the relevant part of the data"""
country_name_to_code = country_names[['COUNTRY_NAME','COUNTRY_ALPHA3_CODE']]
country_name_to_code.COUNTRY_NAME = country_name_to_code.COUNTRY_NAME.str.lower()
"""Drop duplicates since many (or probably all) countries have the same name in several languages"""
country_name_to_code = country_name_to_code[~country_name_to_code.COUNTRY_NAME.duplicated()]
country_name_to_code.head(5)

COUNTRY_NAME COUNTRY_ALPHA3_CODE
0                     andorra                 AND
1  verenigde arabiese emirate                 ARE
2                  afganistan                 AFG
3          antigua en barbuda                 ATG
4                    anguilla                 AIA

In [9]:
clean_df = original_df.merge(country_name_to_code, left_on="country", right_on="COUNTRY_NAME", how="left")
clean_df.drop(["COUNTRY_NAME"], axis=1, inplace=True)

In [10]:
clean_df.head(10)

id  age                                 location  \
0  5749314  NaN                           Moscow, Russia   
1  5751170  NaN                               Bangladesh   
2  5752638  NaN                      Berlin, Deutschland   
3  5754033  NaN                                  Ukraine   
4  5757173  NaN                                      NaN   
5  5766670  NaN              Bangalore, Karnataka, India   
6  5770633  NaN  São Carlos - State of São Paulo, Brazil   
7  5772393  NaN                                      NaN   
8  5775462  NaN                                      NaN   
9  5781104  NaN                                      NaN   

                 creation_date  reputation  up_votes  down_votes      country  \
0   2016-01-05 19:30:46.89 UTC         628        34          69       russia   
1  2016-01-06 07:51:18.563 UTC         412        11           2   bangladesh   
2   2016-01-06 14:11:34.04 UTC          58         4           0  deutschland   
3   2016-01-06 19:20:31.56 UTC          44        96           0      ukraine   
4  2016-01-07 11:20:00.433 UTC          12         0           0          NaN   
5  2016-01-09 12:24:29.703 UTC         386        97           5        india   
6  2016-01-10 18:52:37.347 UTC          84         7           0       brazil   
7  2016-01-11 07:27:41.903 UTC          22         3           0          NaN   
8  2016-01-11 20:06:16.657 UTC         361        19           0          NaN   
9  2016-01-12 21:07:05.793 UTC         299         3           0          NaN   

  COUNTRY_ALPHA3_CODE  
0                 RUS  
1                 BGD  
2                 DEU  
3                 UKR  
4                 NaN  
5                 IND  
6                 BRA  
7                 NaN  
8                 NaN  
9                 NaN

In [11]:
print("Proportion of matches: ", sum(clean_df.COUNTRY_ALPHA3_CODE.notnull())/sum(clean_df.location.notnull()))

Proportion of matches:  0.8697352038731124


The method manages to add a country tag to 86.9% of the users who have informed a country in the first place. Let's then look at the users that weren't matched.

In [12]:
clean_df[clean_df.location.notnull() & clean_df.COUNTRY_ALPHA3_CODE.isnull()]

id  age            location                creation_date  \
32       5876676  NaN  中国广东省Guangzhou Shi  2016-02-03 07:47:09.317 UTC   
34       5882162  NaN              munich  2016-02-04 08:54:15.517 UTC   
49       5924453  NaN            Hobbiton  2016-02-14 06:50:46.743 UTC   
91       6133392  NaN            Shanghai   2016-03-30 07:41:15.61 UTC   
189      6672580  NaN           Melbourne  2016-08-03 13:18:01.843 UTC   
...          ...  ...                 ...                          ...   
6035596  6850818  NaN        Sunnyvale CA   2016-09-19 22:19:35.05 UTC   
6035607  8952195  NaN           NotNeeded  2017-11-16 14:26:13.343 UTC   
6035647  8604749  NaN           Bangalore  2017-09-13 17:09:08.297 UTC   
6035652  5859896  NaN            houbouby   2016-01-30 03:30:18.22 UTC   
6035685  9862136  NaN     Everett, 워싱턴 미국  2018-05-29 04:55:22.243 UTC   

         reputation  up_votes  down_votes             country  \
32               92        86           0  中国广东省guangzhou shi   
34              211        33           0              munich   
49              610        40           5            hobbiton   
91               41         1           0            shanghai   
189            1328       420           1           melbourne   
...             ...       ...         ...                 ...   
6035596          31         0           0        sunnyvale ca   
6035607          31         2           0           notneeded   
6035647          31         0           0           bangalore   
6035652          31         0           0            houbouby   
6035685          31        10           0              워싱턴 미국   

        COUNTRY_ALPHA3_CODE  
32                      NaN  
34                      NaN  
49                      NaN  
91                      NaN  
189                     NaN  
...                     ...  
6035596                 NaN  
6035607                 NaN  
6035647                 NaN  
6035652                 NaN  
6035685                 NaN  

[233629 rows x 9 columns]

Seems like a part of the problem is people giving out only the city or some funny stuff. Let's try addressing this by loading a dataset of city names and trying to match them with the remaining, unmatched data.

In [13]:
world_cities = pd.read_csv("data/worldcities.csv")
display(world_cities)

city     city_ascii      lat       lng        country iso2  \
0          Malishevë      Malisheve  42.4822   20.7458         Kosovo   XK   
1            Prizren        Prizren  42.2139   20.7397         Kosovo   XK   
2        Zubin Potok    Zubin Potok  42.9144   20.6897         Kosovo   XK   
3           Kamenicë       Kamenice  42.5781   21.5803         Kosovo   XK   
4               Viti           Viti  42.3214   21.3583         Kosovo   XK   
...              ...            ...      ...       ...            ...  ...   
12954        Sturgis        Sturgis  44.4111 -103.4975  United States   US   
12955            Tea            Tea  43.4515  -96.8346  United States   US   
12956        Brandon        Brandon  43.5928  -96.5799  United States   US   
12957        Madison        Madison  44.0062  -97.1084  United States   US   
12958  Belle Fourche  Belle Fourche  44.6640 -103.8564  United States   US   

      iso3    admin_name capital  population          id  
0      XKS     Malishevë   admin         NaN  1901597212  
1      XKS       Prizren   admin         NaN  1901360309  
2      XKS   Zubin Potok   admin         NaN  1901608808  
3      XKS      Kamenicë   admin         NaN  1901851592  
4      XKS          Viti   admin         NaN  1901328795  
...    ...           ...     ...         ...         ...  
12954  USA  South Dakota     NaN      6862.0  1840002174  
12955  USA  South Dakota     NaN      5415.0  1840002841  
12956  USA  South Dakota     NaN      9263.0  1840002650  
12957  USA  South Dakota     NaN      6983.0  1840002540  
12958  USA  South Dakota     NaN      5202.0  1840002127  

[12959 rows x 11 columns]

In [14]:
display(world_cities[world_cities.city=="Moscow"])

city city_ascii      lat       lng        country iso2 iso3  \
2220   Moscow     Moscow  55.7522   37.6155         Russia   RU  RUS   
11641  Moscow     Moscow  46.7307 -116.9986  United States   US  USA   

      admin_name  capital  population          id  
2220      Moskva  primary  10452000.0  1643318494  
11641      Idaho      NaN     25579.0  1840019868

There are cities with the same names in many countries. Let's assume that if no country was said, it's probably the largest city.
One could in principle also use the administrative area name here but it doesn't seem as likely that people would put just the name of the administrative area but not the country or city name so let's try the city first.

In [15]:
world_cities = world_cities.sort_values(by="population").drop_duplicates(subset="city",keep="last").sort_index()
display(world_cities[world_cities.city=="Moscow"])

city city_ascii      lat      lng country iso2 iso3 admin_name  \
2220  Moscow     Moscow  55.7522  37.6155  Russia   RU  RUS     Moskva   

      capital  population          id  
2220  primary  10452000.0  1643318494

In [16]:
"""Again, take only interesting parts"""
city_to_country_code = world_cities[['city','iso3']]
city_to_country_code.city = city_to_country_code.city.str.lower()

In [17]:
display(city_to_country_code.head(5))

city iso3
0    malishevë  XKS
1      prizren  XKS
2  zubin potok  XKS
3     kamenicë  XKS
4         viti  XKS

In [18]:
clean_df["city_candidate"] = clean_df["location"].apply(lambda x: x.split(", ")[0].lower() if isinstance(x,str) else np.nan)

In [19]:
display(clean_df.head(5))

id  age             location                creation_date  reputation  \
0  5749314  NaN       Moscow, Russia   2016-01-05 19:30:46.89 UTC         628   
1  5751170  NaN           Bangladesh  2016-01-06 07:51:18.563 UTC         412   
2  5752638  NaN  Berlin, Deutschland   2016-01-06 14:11:34.04 UTC          58   
3  5754033  NaN              Ukraine   2016-01-06 19:20:31.56 UTC          44   
4  5757173  NaN                  NaN  2016-01-07 11:20:00.433 UTC          12   

   up_votes  down_votes      country COUNTRY_ALPHA3_CODE city_candidate  
0        34          69       russia                 RUS         moscow  
1        11           2   bangladesh                 BGD     bangladesh  
2         4           0  deutschland                 DEU         berlin  
3        96           0      ukraine                 UKR        ukraine  
4         0           0          NaN                 NaN            NaN

In [20]:
clean_df = clean_df.merge(city_to_country_code, left_on="city_candidate", right_on="city", how="left")
clean_df["COUNTRY_ALPHA3_CODE"][clean_df["COUNTRY_ALPHA3_CODE"].isnull()] = clean_df["iso3"][clean_df["COUNTRY_ALPHA3_CODE"].isnull()]

In [21]:
print("Proportion of matched: ", sum(clean_df.COUNTRY_ALPHA3_CODE.notnull())/sum(clean_df.location.notnull()))

Proportion of matched:  0.9237611486309442


Seems like using the city name brings us up to 92.3% of matches.

In [22]:
"""Clean out some useless fields"""
clean_df.drop(["city_candidate", "city", "iso3","country"],axis=1, inplace=True)

In [23]:
"""See the remaining users that are still unmatched. Also prepare a lowercase location column for further analysis."""
clean_df["location_lc"] = clean_df.location.str.lower()
remaining = clean_df[clean_df.location.notnull()][clean_df.COUNTRY_ALPHA3_CODE.isnull()]
display(remaining[100:110])

id  age               location                creation_date  \
3147  7115354  NaN                     UK  2016-11-04 13:50:40.537 UTC   
3153  7140620  NaN                   Mahé   2016-11-10 09:57:51.26 UTC   
3208  7479558  NaN  Shipston-on-Stour, UK  2017-01-27 15:44:07.993 UTC   
3237  7652266  NaN     KAIST, South Korea  2017-03-03 08:30:54.243 UTC   
3251  7783970  NaN          日本Saitama-ken  2017-03-29 06:13:11.537 UTC   
3267  7885186  NaN                  world   2017-04-18 16:01:59.25 UTC   
3285  8001672  NaN             대한민국 Seoul  2017-05-12 08:04:23.997 UTC   
3366  8710344  NaN                 Kanata   2017-10-02 19:17:11.91 UTC   
3543  5812981  NaN               location  2016-01-19 22:36:39.597 UTC   
3596  6051830  NaN                 Europe  2016-03-11 20:16:45.963 UTC   

      reputation  up_votes  down_votes COUNTRY_ALPHA3_CODE  \
3147         351       151           2                 NaN   
3153          30        30           0                 NaN   
3208         157         6           0                 NaN   
3237          92        49           0                 NaN   
3251         113         0           0                 NaN   
3267          45         5           0                 NaN   
3285          23         1           0                 NaN   
3366        2206       389          39                 NaN   
3543         825       276           6                 NaN   
3596         289       200           1                 NaN   

                location_lc  
3147                     uk  
3153                   mahé  
3208  shipston-on-stour, uk  
3237     kaist, south korea  
3251          日本saitama-ken  
3267                  world  
3285             대한민국 seoul  
3366                 kanata  
3543               location  
3596                 europe

Seems like the remaining problem is in part that the format isn't necessarily
always city, admin area, country but sometimes some other ordering. Also, for instance UK is a common abbreviation for the United Kingdom but it wasn't in the country dataset. A regex search for the rest seems like a good idea.

In [24]:
country_name_to_code_2 = country_name_to_code.append([{"COUNTRY_NAME" : "uk", "COUNTRY_ALPHA3_CODE" : "GBR"},
                                                     {"COUNTRY_NAME" : "usa", "COUNTRY_ALPHA3_CODE" : "USA"},
                                                     {"COUNTRY_NAME" : "south korea", "COUNTRY_ALPHA3_CODE" : "KOR"}] ,ignore_index=True)

In [25]:
i = 0
for idx, row in country_name_to_code_2.iterrows():
    country_name,iso3 = row[0],row[1]
    matching = remaining.location_lc.str.contains("\\b"+country_name+"\\b")
    remaining.COUNTRY_ALPHA3_CODE[matching] = iso3
    if i % 1000 == 0:
        print(idx/len(country_name_to_code)*100, "% calculated")
    i += 1

0.0 % calculated
7.471049682480388 % calculated
14.942099364960775 % calculated
22.413149047441166 % calculated
29.88419872992155 % calculated
37.35524841240194 % calculated
44.82629809488233 % calculated
52.297347777362724 % calculated
59.7683974598431 % calculated
67.2394471423235 % calculated
74.71049682480388 % calculated
82.18154650728428 % calculated
89.65259618976467 % calculated
97.12364587224505 % calculated


In [26]:
clean_df = clean_df.merge(remaining.drop(["id","location","creation_date","reputation","up_votes","down_votes","location_lc","age"],axis=1), left_index=True, right_index=True, how="left")
clean_df.COUNTRY_ALPHA3_CODE_x[clean_df.COUNTRY_ALPHA3_CODE_x.isnull()] = clean_df.COUNTRY_ALPHA3_CODE_y[clean_df.COUNTRY_ALPHA3_CODE_x.isnull()]

In [27]:
clean_df.drop(["COUNTRY_ALPHA3_CODE_y"], axis=1, inplace=True)
clean_df.rename(columns={"COUNTRY_ALPHA3_CODE_x" : "country_iso3"}, inplace=True)

In [28]:
print("Proportion of matched: ", sum(clean_df.country_iso3.notnull())/sum(clean_df.location.notnull()))

Proportion of matched:  0.9474566236164214


Using the regex method brings the results already to 94.75 percent. Let's look at the remaining issues.

In [29]:
clean_df[clean_df.location.notnull() & clean_df.country_iso3.isnull()][160:180]

id  age                            location  \
8380   9333556  NaN                             Czechia   
8409   9719937  NaN                           Worldwide   
8538   5991722  NaN   Hranice na Moravě, Hranice, Česko   
8622   6292000  NaN                           Somewhere   
8732   6783798  NaN                         Schweinfurt   
8800   7165275  NaN                             Vermont   
8901   7771928  NaN                            Bangaore   
8924   7878102  NaN          Nagrota Bypass Road, Jammu   
8931   7898657  NaN               中国河北省Shijiazhuang Shi   
8947   7974340  NaN                                U.S.   
9004   8377142  NaN                       Beroun, Česko   
9007   8398531  NaN                               Korea   
9031   8580067  NaN               Lviv, Lviv Oblast, UA   
9074   8998175  NaN                               @home   
9180   9819333  NaN                               Noida   
9215  10306949  NaN                                  US   
9291   5971239  NaN  San Francisco Bay Area, California   
9294   6004645  NaN                     Wisconsin, U.S.   
9425   6589588  NaN         Charlotte · Bremen · Taipei   
9491   6820646  NaN          Somewhere over the rainbow   

                    creation_date  reputation  up_votes  down_votes  \
8380  2018-02-08 14:04:47.567 UTC          30         0           0   
8409   2018-04-30 06:38:27.79 UTC          30        47           0   
8538   2016-02-27 22:37:37.31 UTC         461        27          17   
8622  2016-05-04 17:01:26.607 UTC         308       693           0   
8732   2016-09-01 14:29:22.53 UTC        1061       129          26   
8800  2016-11-16 03:00:26.087 UTC        1943        60          12   
8901  2017-03-27 04:26:53.803 UTC          37         0           0   
8924   2017-04-17 10:18:59.57 UTC          61         8           0   
8931  2017-04-21 00:05:31.497 UTC         431         4           3   
8947  2017-05-06 20:41:23.647 UTC         126        12           0   
9004  2017-07-27 15:14:26.397 UTC          52         3           0   
9007  2017-08-01 08:41:32.603 UTC          15         1           0   
9031   2017-09-08 12:36:22.56 UTC          18         0           0   
9074  2017-11-23 20:06:40.347 UTC         149        15           2   
9180   2018-05-20 15:10:25.88 UTC         802        10           0   
9215  2018-09-02 15:19:39.513 UTC         111         5           0   
9291   2016-02-23 22:15:53.26 UTC        1184        45           1   
9294   2016-03-01 20:53:43.71 UTC          24         6           0   
9425   2016-07-14 13:04:20.55 UTC         328         1           0   
9491  2016-09-12 01:43:37.047 UTC         154        61           0   

     country_iso3                         location_lc  
8380          NaN                             czechia  
8409          NaN                           worldwide  
8538          NaN   hranice na moravě, hranice, česko  
8622          NaN                           somewhere  
8732          NaN                         schweinfurt  
8800          NaN                             vermont  
8901          NaN                            bangaore  
8924          NaN          nagrota bypass road, jammu  
8931          NaN               中国河北省shijiazhuang shi  
8947          NaN                                u.s.  
9004          NaN                       beroun, česko  
9007          NaN                               korea  
9031          NaN               lviv, lviv oblast, ua  
9074          NaN                               @home  
9180          NaN                               noida  
9215          NaN                                  us  
9291          NaN  san francisco bay area, california  
9294          NaN                     wisconsin, u.s.  
9425          NaN         charlotte · bremen · taipei  
9491          NaN          somewhere over the rainbow

Seems like the remaining issues are mostly people giving out only the city or administrative area name in some format that wasn't considered before. Let's try to fix this with a similar regex matching for both.

In [30]:
world_cities_2 = pd.read_csv("data/worldcities.csv")
world_cities_2 = world_cities.sort_values(by="population").drop_duplicates(subset="admin_name",keep="last").sort_index()
admin_to_country_code = world_cities_2[['admin_name','iso3']]
admin_to_country_code.admin_name = admin_to_country_code.admin_name.str.lower()

In [32]:
"""Weird thing: NaN value in one of the admin names"""
remaining = clean_df[clean_df.location_lc.notnull()][clean_df.country_iso3.isnull()]
i = 0
for idx, row in admin_to_country_code.iterrows():
    admin_name,iso3 = row[0],row[1]
    matching = remaining.location_lc.str.contains("\\b"+str(admin_name)+"\\b")
    remaining.country_iso3[matching] = iso3
    if i % 1000 == 0:
        print(i/len(admin_to_country_code)*100, "% calculated")
    i += 1

0.0 % calculated
25.73340195573855 % calculated
51.4668039114771 % calculated
77.20020586721564 % calculated


In [33]:
remaining.rename(columns={"country_iso3" : "rem_iso3"}, inplace=True)

In [34]:
clean_df = clean_df.merge(remaining.drop(["id","location","creation_date","reputation","up_votes","down_votes","location_lc","age"],axis=1), left_index=True, right_index=True, how="left")
clean_df.country_iso3[clean_df.country_iso3.isnull()] = clean_df.rem_iso3[clean_df.country_iso3.isnull()]
clean_df.drop(["rem_iso3"], axis=1, inplace=True)

In [35]:
remaining = clean_df[clean_df.location_lc.notnull() & clean_df.country_iso3.isnull()]
i = 0
for idx, row in city_to_country_code.iterrows():
    city_name,iso3 = row[0],row[1]
    matching = remaining.location_lc.str.contains("\\b"+city_name+"\\b")
    remaining.country_iso3[matching] = iso3
    if i % 1000 == 0:
        print(i/len(city_to_country_code)*100, "% calculated")
    i += 1

0.0 % calculated
8.634832915983075 % calculated
17.26966583196615 % calculated
25.90449874794923 % calculated
34.5393316639323 % calculated
43.17416457991538 % calculated
51.80899749589846 % calculated
60.44383041188153 % calculated
69.0786633278646 % calculated
77.71349624384769 % calculated
86.34832915983075 % calculated
94.98316207581384 % calculated


In [36]:
remaining.rename(columns={"country_iso3" : "rem_iso3"}, inplace=True)

In [37]:
clean_df = clean_df.merge(remaining.drop(["id","location","creation_date","reputation","up_votes","down_votes","location_lc","age"],axis=1), left_index=True, right_index=True, how="left")
clean_df.country_iso3[clean_df.country_iso3.isnull()] = clean_df.rem_iso3[clean_df.country_iso3.isnull()]
clean_df.drop(["rem_iso3"], axis=1, inplace=True)

In [38]:
display(admin_to_country_code[[not isinstance(x,str) for x in admin_to_country_code.admin_name]])

admin_name iso3
7818        NaN  JEY

Funny fact: before we needed to have str(admin_name) in the code because there's a place called Nan in Thailand. 

In [39]:
sum(clean_df.country_iso3.notnull())/sum(clean_df.location.notnull())

0.956600201369446

With the final method we get 95.66% of country matches to locations.

In [40]:
clean_df.drop(["location","location_lc"],axis=1,inplace=True)

In [41]:
clean_df.to_csv("data/stackoverflow_users_2016-2019_notext_clean.csv")

In [42]:
original_df[clean_df.country_iso3.isnull() & original_df.location.notnull()]

id  age                              location  \
32        5876676  NaN                    中国广东省Guangzhou Shi   
49        5924453  NaN                              Hobbiton   
244       7008814  NaN                                  mars   
425       7878292  NaN                          Arandjelovac   
435       7945249  NaN                             127.0.0.1   
...           ...  ...                                   ...   
6035679   6142801  NaN             New Jersey, United States   
6035680   9703547  NaN  Yokohama, Kanagawa Prefecture, Japan   
6035682  11258913  NaN                      Mississippi, USA   
6035685   9862136  NaN                       Everett, 워싱턴 미국   
6035687  11636649  NaN          Nagercoil, Tamil Nadu, India   

                       creation_date  reputation  up_votes  down_votes  \
32       2016-02-03 07:47:09.317 UTC          92        86           0   
49       2016-02-14 06:50:46.743 UTC         610        40           5   
244      2016-10-12 15:28:03.683 UTC          15        68           0   
425        2017-04-17 11:10:19.7 UTC          18         3           0   
435       2017-04-30 22:50:32.76 UTC         154         7           0   
...                              ...         ...       ...         ...   
6035679  2016-04-01 01:57:26.287 UTC          31         0           0   
6035680  2018-04-26 10:47:03.497 UTC          31         2           0   
6035682  2019-03-26 07:21:08.673 UTC          31        24           0   
6035685  2018-05-29 04:55:22.243 UTC          31        10           0   
6035687  2019-06-12 12:03:40.373 UTC          31        28           0   

                    country  
32       中国广东省guangzhou shi  
49                 hobbiton  
244                    mars  
425            arandjelovac  
435               127.0.0.1  
...                     ...  
6035679       united states  
6035680               japan  
6035682                 usa  
6035685              워싱턴 미국  
6035687               india  

[1228364 rows x 8 columns]

In [43]:
original_df[original_df.location.notnull() & clean_df.country_iso3.isnull()][100:130]

id  age                                           location  \
6163   8644091  NaN                                         Babu Nagar   
6174   8813176  NaN                                     Mariana Trench   
6180   8867531  NaN                                            Czechia   
6203   9015220  NaN                                             Korean   
6211   9061350  NaN                                        Malu abrupt   
6262   9537794  NaN                                         Everywhere   
6309   9975719  NaN                                                 DC   
6327  10211923  NaN                                          HumanLand   
6356   5789789  NaN                                        GetPosition   
6446   6176383  NaN                                          Milky-Way   
6508   6440619  NaN                                          Somewhere   
6575   6706654  NaN                                           Internet   
6603   6821443  NaN                                              中国上海市   
6689   7358586  NaN                                           The moon   
6788   7924725  NaN                                             Europe   
6819   8182375  NaN                                             Ismail   
6860   8496457  NaN                                             Europa   
6951   9188006  NaN                                          The earth   
6975   9394708  NaN                                           ghansoli   
7026   9799599  NaN                                              HKUST   
7055  10205276  NaN                                    Praha 10, Česko   
7079   5774750  NaN                                         Strickland   
7084   5793033  NaN                                             States   
7090   5813228  NaN                                        Cyber Space   
7213   6335837  NaN                                     中国Shanghai Shi   
7278   6630464  NaN                                       The Internet   
7286   6648456  NaN                                             Europe   
7312   6749539  NaN  In the desert eating dessert contemplating whe...   
7407   7296192  NaN  Universe, Virgo Supercluster, Local Group, Mil...   
7470   7663858  NaN                                  中国上海市Shanghai Shi   

                    creation_date  reputation  up_votes  down_votes  \
6163  2017-09-20 19:28:47.367 UTC         429       166          19   
6174   2017-10-22 03:32:06.66 UTC          93        10           0   
6180  2017-11-01 10:38:22.473 UTC          29        18           0   
6203  2017-11-27 14:22:04.717 UTC          55         9           0   
6211   2017-12-06 11:10:36.39 UTC          36         0           0   
6262  2018-03-23 00:48:57.613 UTC          55         2           0   
6309  2018-06-21 23:43:00.367 UTC          30         2           0   
6327   2018-08-11 10:33:01.33 UTC         632        74          10   
6356  2016-01-14 12:56:40.593 UTC          23         0           0   
6446   2016-04-08 09:17:07.11 UTC          96         5           0   
6508  2016-06-08 13:31:30.297 UTC          38         9           0   
6575  2016-08-11 19:49:52.637 UTC          65        17           0   
6603  2016-09-12 07:37:04.213 UTC         106         0           0   
6689  2016-12-30 16:45:07.183 UTC         166        27           4   
6788  2017-04-26 10:16:26.937 UTC          58         5           0   
6819  2017-06-19 09:59:39.717 UTC          36         9           0   
6860   2017-08-21 17:29:32.95 UTC         427         6          14   
6951   2018-01-08 12:48:32.05 UTC          84        17           0   
6975  2018-02-22 06:18:41.527 UTC          14         0           0   
7026  2018-05-16 10:43:15.847 UTC          43         2           0   
7055  2018-08-09 19:57:13.913 UTC         365        45           1   
7079  2016-01-11 16:52:58.473 UTC         857        27           0   
7084   2016-01-15 04:46:37.41 UTC        1242       118         656   
7090   2016-0

Looks that most of the unmatched entries are just people trying to be funny. "Earth" and "Moon" are quite popular locations, for instance.

In [44]:
world_cities[world_cities.city.str.lower().str.contains("\\bada\\b")]

city city_ascii      lat      lng country iso2 iso3 admin_name capital  \
2010  Ada        Ada  45.8027  20.1285  Serbia   RS  SRB        Ada   admin   

      population          id  
2010         NaN  1688687750

There's also a city called Ada in Serbia. :)